pkg/app/AppListingService.go

```diff
@@ -769,6 +802,91 @@ func (impl AppListingServiceImpl) FetchAppStageStatus(appId int, appType int) ([
 	appStageStatuses, err := impl.appListingRepository.FetchAppStageStatus(appId, appType)
 	return appStageStatuses, err
 }
+func (impl AppListingServiceImpl) fetchCiArtifactAndGitTriggersMapping(artifacts []*CiArtifactWithParentArtifact, ctx context.Context) (CiArtifactAndGitCommitsMap map[int][]string, err error) {
+	newCtx, span := otel.Tracer("envOverrideRepository").Start(ctx, "fetchCiArtifactAndGitTriggersMapping")
+	defer span.End()
+	ciArtifactIds, artifactChildParentMap, err := mapCiArtifactsWithChild(artifacts)
+	if err != nil {
+		// Log the error along with the artifact IDs that caused it.
+		impl.Logger.Errorw("error in getting the set of ciArtifactIds", "ArtifactIds", ciArtifactIds, "err", err)
+		return nil, err // Return nil map and the encountered error.
+	}
+
+	// Retrieve workflows associated with the artifact IDs, handling any potential error.
+	artifactsWithGitTriggers, err := impl.ciWorkflowRepository.FindGitTriggersByArtifactIds(ciArtifactIds, newCtx)
+	if err != nil {
+		// Log the error along with the artifact IDs that caused it.
+		impl.Logger.Errorw("error retrieving GitTriggers of the  CiWorkflows", "ciArtifactIds", ciArtifactIds, "err", err)
+		return nil, err // Return nil map and the encountered error.
+	}
+
+	gitCommitsWithChildArtifactMap, err := mapArtifactToGitCommits(artifactsWithGitTriggers, artifactChildParentMap)
+	if err != nil {
+		// Log the error along with the artifact child IDs .
+		impl.Logger.Errorw("error retrieving GitCommits of these ids", "CiArtifactIds", ciArtifactIds, "err", err)
+		return nil, err
+	}
+
+	return gitCommitsWithChildArtifactMap, nil
+}
+
+func mapCiArtifactsWithChild(artifacts []*CiArtifactWithParentArtifact) (ciArtifactIds []int, artifactChildParentMap map[int]int, err error) {
+	artifactChildParentMap = make(map[int]int)
+
+	for _, artifact := range artifacts {
+		// Mapping the current artifact to its parent, or to itself if it has no parent.
+		if artifact.ParentCiArtifact > 0 {
+			artifactChildParentMap[artifact.CiArtifactId] = artifact.ParentCiArtifact
+		} else {
+			artifactChildParentMap[artifact.CiArtifactId] = artifact.CiArtifactId
+		}
+
+		// Ensure uniqueness of artifact IDs in the slice.
+		if ciArtifactIds == nil {
+			if _, ok := artifactChildParentMap[artifact.CiArtifactId]; ok {
+				ciArtifactIds = []int{artifactChildParentMap[artifact.CiArtifactId]}
+			}
+		} else if !slices.Contains(ciArtifactIds, artifactChildParentMap[artifact.CiArtifactId]) {
```

Comment:
`Contains` is a linear search, invoking inside `for` loop will  make complexity exponential.

Reply:
resolved using map